File: app/src/main/java/com/travlog/android/apps/ui/activities/SignInActivity.java
package com.travlog.android.apps.ui.activities;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.util.Pair;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import com.facebook.login.widget.LoginButton;
import com.google.android.gms.auth.api.Auth;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.SignInButton;
import com.google.android.gms.common.api.GoogleApiClient;
import com.jakewharton.rxbinding2.view.RxView;
import com.jakewharton.rxbinding2.widget.RxTextView;
import com.travlog.android.apps.R;
import com.travlog.android.apps.libs.BaseActivity;
import com.travlog.android.apps.libs.qualifiers.RequiresActivityViewModel;
import com.travlog.android.apps.viewmodels.SignInViewModel;

import butterknife.BindView;
import io.reactivex.android.schedulers.AndroidSchedulers;
import timber.log.Timber;

import static com.travlog.android.apps.libs.ActivityRequestCodes.SIGN_IN_WITH_GOOGLE;
import static com.travlog.android.apps.libs.ActivityRequestCodes.SIGN_UP_FLOW;

@RequiresActivityViewModel(SignInViewModel.class)
public class SignInActivity extends BaseActivity<SignInViewModel> {

    @BindView(R.id.close_button)
    View closeButton;
    @BindView(R.id.sign_in_with_facebook_button)
    LoginButton signInFacebookButton;
    @BindView(R.id.sign_in_with_google_button)
    SignInButton signInGoogleButton;
    @BindView(R.id.email)
    EditText emailEditText;
    @BindView(R.id.password)
    EditText passwordEditText;
    @BindView(R.id.sign_in_button)
    Button signInButton;
    @BindView(R.id.forgot_password_button)
    Button forgotPasswordButton;
    @BindView(R.id.sign_up_button)
    Button signUpButton;

    GoogleApiClient googleApiClient;

    @SuppressLint("CheckResult")
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.a_sign_in);

        signInFacebookButton.setReadPermissions("email", "public_profile");

        final GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestIdToken(getString(R.string.default_web_client_id))
                .requestEmail()
                .build();

        googleApiClient = new GoogleApiClient.Builder(this)
                .enableAutoManage(this,
                        connectionResult -> Timber.d("onConnectionFailed: %s", connectionResult.getErrorMessage()))
                .addApi(Auth.GOOGLE_SIGN_IN_API, gso)
                .build();

        RxView.clicks(closeButton)
                .compose(bindToLifecycle())
                .subscribe(__ -> back());

        RxView.clicks(signInGoogleButton)
                .compose(bindToLifecycle())
                .subscribe(__ -> startSignInWithGoogleActivity());

        RxTextView.textChanges(emailEditText)
                .map(CharSequence::toString)
                .compose(bindToLifecycle())
                .subscribe(viewModel::email);

        RxTextView.textChanges(passwordEditText)
                .map(CharSequence::toString)
                .compose(bindToLifecycle())
                .subscribe(viewModel::password);

        RxView.clicks(signInButton)
                .compose(bindToLifecycle())
                .subscribe(__ -> viewModel.signInClick());

        RxView.clicks(signUpButton)
                .compose(bindToLifecycle())
                .subscribe(__ -> startSignUp());

        viewModel.outputs.signInSuccess()
                .observeOn(AndroidSchedulers.mainThread())
                .compose(bindToLifecycle())
                .subscribe(__ -> this.back());

        viewModel.outputs.setSignInButtonEnabled()
                .compose(bindToLifecycle())
                .subscribe(this::setSignInButtonEnabled);
    }

    private void startSignInWithGoogleActivity() {
        final Intent intent = Auth.GoogleSignInApi.getSignInIntent(googleApiClient);
        startActivityForResult(intent, SIGN_IN_WITH_GOOGLE);
    }

    private void startSignUp() {
        final Intent intent = new Intent(this, SignUpActivity.class);
        startActivityForResult(intent, SIGN_UP_FLOW);
    }

    private void setSignInButtonEnabled(final boolean enabled) {
        signInButton.setEnabled(enabled);
    }

    @Nullable
    @Override
    protected Pair<Integer, Integer> exitTransition() {
        return Pair.create(android.R.anim.fade_in, R.anim.slide_out_bottom);
    }
}

File: app/src/main/java/com/travlog/android/apps/ui/activities/SignInActivity.kt
package com.travlog.android.apps.ui.activities

import android.content.Intent
import android.os.Bundle
import android.util.Pair
import com.google.android.gms.auth.api.Auth
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.Scopes
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.common.api.Scope
import com.jakewharton.rxbinding3.view.clicks
import com.jakewharton.rxbinding3.widget.textChanges
import com.travlog.android.apps.R
import com.travlog.android.apps.libs.ActivityRequestCodes.SIGN_IN_WITH_GOOGLE
import com.travlog.android.apps.libs.ActivityRequestCodes.SIGN_UP_FLOW
import com.travlog.android.apps.libs.BaseActivity
import com.travlog.android.apps.viewmodels.SignInViewModel
import io.reactivex.android.schedulers.AndroidSchedulers
import kotlinx.android.synthetic.main.a_sign_in.*
import timber.log.Timber

class SignInActivity : BaseActivity<SignInViewModel>() {

    private var googleApiClient: GoogleApiClient? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.a_sign_in)

        this.sign_in_with_facebook_button.setReadPermissions("email", "public_profile")

        GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestScopes(Scope(Scopes.EMAIL))
                .requestIdToken(getString(R.string.default_web_client_id))
                .build().let {
//                    googleApiClient = GoogleApiClient.Builder(this)
//                            .enableAutoManage(this) { Timber.d("onConnectionFailed: %s", it.errorMessage) }
//                            .addApi(Auth.GOOGLE_SIGN_IN_API, it)
//                            .build()
                }

        close_button.clicks()
                .compose(bindToLifecycle())
                .subscribe { this.back() }

        sign_in_with_google_button.clicks()
                .compose(bindToLifecycle())
                .subscribe { this.startSignInWithGoogleActivity() }

        viewModel?.apply {
            login_id.textChanges()
                    .map { it.toString() }
                    .compose(bindToLifecycle())
                    .subscribe(inputs::loginId)

            password.textChanges()
                    .map { it.toString() }
                    .compose(bindToLifecycle())
                    .subscribe(inputs::password)

            sign_in_button.clicks()
                    .compose(bindToLifecycle())
                    .subscribe { inputs.signInClick() }

            sign_up_button.clicks()
                    .compose(bindToLifecycle())
                    .subscribe { startSignUp() }

            addDisposable(
                    outputs.startMainActivity()
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribe { this@SignInActivity.startMainActivity() }
            )

            setSignInButtonEnabled()
                    .compose(bindToLifecycle())
                    .subscribe { setSignInButtonEnabled(it) }
        }
    }

    private fun startMainActivity() =
            Intent(this, MainActivity::class.java).let {
                startActivity(it)
                back()
            }

    private fun startSignInWithGoogleActivity() =
            Auth.GoogleSignInApi.getSignInIntent(googleApiClient).let {
                startActivityForResult(it, SIGN_IN_WITH_GOOGLE)
            }

    private fun startSignUp() =
            Intent(this, SignUpActivity::class.java).let {
                startActivityForResult(it, SIGN_UP_FLOW)
            }

    private fun setSignInButtonEnabled(enabled: Boolean) {
        this.sign_in_button.isEnabled = enabled
    }

    override fun exitTransition(): Pair<Int, Int>? =
            Pair.create(android.R.anim.fade_in, R.anim.slide_out_bottom)
}
